src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -122,21 +122,26 @@ oop SystemDictionary::java_platform_loader() {
 void SystemDictionary::compute_java_loaders(TRAPS) {
   JavaValue result(T_OBJECT);
   InstanceKlass* class_loader_klass = vmClasses::ClassLoader_klass();
-  JavaCalls::call_static(&result,
-                         class_loader_klass,
-                         vmSymbols::getSystemClassLoader_name(),
-                         vmSymbols::void_classloader_signature(),
-                         CHECK);
 
-  _java_system_loader = OopHandle(Universe::vm_global(), result.get_oop());
+  if (_java_system_loader.is_empty()) {
+    JavaCalls::call_static(&result,
+                           class_loader_klass,
+                           vmSymbols::getSystemClassLoader_name(),
+                           vmSymbols::void_classloader_signature(),
+                           CHECK);
 
-  JavaCalls::call_static(&result,
-                         class_loader_klass,
-                         vmSymbols::getPlatformClassLoader_name(),
-                         vmSymbols::void_classloader_signature(),
-                         CHECK);
+    _java_system_loader = OopHandle(Universe::vm_global(), result.get_oop());
+  }
```

Comment:
To check that the we correctly restored _java_system_loader from CDS, I think we should move the Java upcall to a worker function and add some asserts:

```
if (_java_system_loader.is_empty()) {
    _java_system_loader = get_system_classloader_impl();
} else {
   // It must have been restored from the archived module graph
   assert(UseSharedSpaces, "must be");
   assert(MetaspaceShared::use_full_module_graph(). "must be");
   assert(_java_system_loader.resolve() == get_system_classloader_impl(), "must be");
}
```

Reply:
Done.